jdat_notebooks/optimal_extraction/Spectral Extraction.ipynb

```diff
@@ -0,0 +1,1746 @@
+    "        print(\"Processing file {} of {}: {}\".format(i+1, len(file_list), fitsfile))\n",
+    "        dmodel = ImageModel(fitsfile)\n",
+    "        spec2d = dmodel.data\n",
+    "        wht2d = dmodel.wht\n",
+    "        \n",
+    "        k_slice = batch_kernel_slice(spec2d)\n",
+    "        k_model = batch_fit_extraction_kernel(k_slice)\n",
+    "        trace = batch_fit_trace_centers(spec2d, k_model)\n",
+    "        spectrum = batch_extract_spectrum(spec2d, trace, k_model, wht2d)\n",
+    "        \n",
+    "        ny, nx = spec2d.shape\n",
+    "        y2d, x2d = np.mgrid[:ny, :nx]\n",
+    "        wavelength = batch_wavelength_from_wcs(dmodel, x2d, y2d)\n",
+    "        \n",
+    "        bbox = [0, 0, nx-1, ny-1]\n",
+    "        \n",
+    "        outfile = fitsfile.replace('s2d.fits', 'x1d_optimal')\n",
+    "        \n",
+    "        batch_plot_output(spec2d, bbox, k_slice, k_model,\n",
+    "                    wavelength, spectrum, \n",
+    "                    outfile+'.png')\n",
+    "        batch_save_extracted_spectrum(outfile+'.fits', wavelength, spectrum)\n",
+    "        "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Run on example dataset\n",
+    "\n",
+    "Take particular note of any spectrum which produces a warning during fitting - these are likely to be good candidates for interactive reprocessing."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "s2d_files = glob(os.path.join('s2d_files', '*s2d.fits'))\n",
+    "batch_optimal_extraction(s2d_files)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "---"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Appendix B: WebbPSF"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Instead of using a PSF template, we can generate a PSF directly from the instrument model with [WebbPSF](https://webbpsf.readthedocs.io/en/stable/index.html). Currently, only the F110W and F140X imaging filters are supported, but we'll walk through the process anyway for whenever more filters become available.\n",
+    "\n",
+    "The primary function of WebbPSF is to produce imaging PSFs; however, it *can* generate a set of monochromatic PSFs, which we can combine."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "***to be added: WebbPSF installation / dataset instructions*** "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Instrument properties\n",
+    "See the WebbPSF documentation for a full list of instrument settings."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "instrument = NIRSpec()\n",
```

Comment:
I'm still getting the same error as before: https://github.com/spacetelescope/dat_pyinthesky/pull/105#pullrequestreview-439483053